### PR TITLE
Improve skills and fix gameplay bugs

### LINKED ISSUE #3137: Make assignment tactic "sticky"
### Description of the task

We can try to make it tactic assignment "sticky" to prevent oscillations.

Some work was previously done in #2342 

### Acceptance criteria

- [ ] Write a unit test (count number of assignments without "sticky"-ness and number of assignments with "sticky" -> we expect it to be lower)

### PR DESCRIPTION
**Pull request by williamckha**
### Description

- "Sticky" tactic assignment: if a tactic was previously assigned to a robot, its assignment cost for the robot is decreased so that it will likely be assigned to the robot again
- Find best shot for `ShootSkill` by sampling nearby shot origin points; the ball is dribbled to the optimal shot origin point and then shot from there
- Tuned skill feasibility scoring

TODO: fix PassSkill and wait to pass until receiver is in position

### Resolved Issues

Resolves #3137 

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/ai/evaluation/calc_best_shot.cpp [hunk 1]
```diff
@@ -86,7 +89,8 @@ std::optional<Shot> calcBestShotOnGoal(const Segment &goal_post, const Point &sh
     Point shot_point = (top_point - bottom_point) / 2 + bottom_point;
 
     return std::make_optional(
-        Shot(shot_point, Angle::fromDegrees(biggest_angle_seg.getDeltaInDegrees())));
+        Shot(shot_origin, shot_point,
```
Comment (Bvasilchikov) [inline src/software/ai/evaluation/calc_best_shot.cpp]: nit: 
```suggestion
        Shot(shot_origin, shot_target,
```

File: src/software/ai/evaluation/calc_best_shot.cpp [hunk 2]
```diff
@@ -147,3 +151,44 @@ std::optional<Shot> calcBestShotOnGoal(const Field &field, const Team &friendly_
...
+std::optional<Shot> findBestShotOnGoal(const Field &field, const Team &friendly_team,
+                                       const Team &enemy_team,
+                                       const Point &starting_point, TeamType goal,
+                                       const std::vector<Robot> &robots_to_ignore,
+                                       double radius)
+{
+    // Number of shot origin points to sample to the left and right of the starting point
+    static constexpr int NUM_SAMPLE_POINTS = 5;
+    // Spacing between sample points in metres
+    static constexpr double SAMPLE_POINTS_SPACING_M = 0.1;
```
Comment (Bvasilchikov) [inline src/software/ai/evaluation/calc_best_shot.cpp]: nit: should move these to constants file

File: src/software/ai/evaluation/calc_best_shot.cpp [hunk 3]
```diff
@@ -147,3 +151,44 @@ std::optional<Shot> calcBestShotOnGoal(const Field &field, const Team &friendly_
...
+
+        std::optional<Shot> shot =
+            calcBestShotOnGoal(field, friendly_team, enemy_team, shot_origin, goal,
+                               robots_to_ignore, radius);
+
+        if (shot && (!best_shot || shot->getOpenAngle() > best_shot->getOpenAngle()))
+        {
+            best_shot = shot;
+        }
+    }
```
Comment (Bvasilchikov) [inline src/software/ai/evaluation/calc_best_shot.cpp]: Suggestion: I can recommend trying performance of using `std::transform` and then calling `std::sort` on the list which would looker cleaner potentially , which should be O(n*log(n)), as the sort dominates - call me out if I'm wrong here on the analysis.

That would give you all the sampled shots ranked, which might be useful information later in the future.

File: src/software/ai/evaluation/calc_best_shot.h [hunk 4]
```diff
@@ -58,3 +58,36 @@ std::optional<Shot> calcBestShotOnGoal(const Field &field, const Team &friendly_
...
+ * @param starting_point The point to begin sampling shot origin points from
+ * @param goal The goal to shoot at
+ * @param robots_to_ignore The robots to ignore
+ * @param radius The radius for the robot obstacles
+ *
+ * @return the best best origin to shoot from, target to shoot at, and the largest open
+ * angle interval for the shot (this is the total angle between the obstacles on either
+ * side of the shot vector). If no shot can be found, returns std::nullopt
+ */
+std::optional<Shot> findBestShotOnGoal(const Field &field, const Team &friendly_team,
```
Comment (Bvasilchikov) [inline src/software/ai/evaluation/calc_best_shot.h]:
```suggestion
std::optional<Shot> sampleForBestShotOnGoal(const Field &field, const Team &friendly_team,
```

File: src/software/ai/evaluation/scoring/skills/feasibility_visitor.cpp [hunk 5]
```diff
@@ -32,44 +32,30 @@ void FeasibilityVisitor::visit(const KeepAwaySkill& skill)
...
-    if (best_pass.rating > min_score)
-    {
-        // TODO(arun): should we normalize this?
-        current_feasibility_ = best_pass.rating;
-        return;
-    }
-
-    current_feasibility_ = 0;
-    return;
+    current_feasibility_ = normalizeValueToRange(best_pass.rating, 0.0, 0.7, 0.0, 1.0);
```
Comment (Bvasilchikov) [inline src/software/ai/evaluation/scoring/skills/feasibility_visitor.cpp:36]: What is the purpose or why is this change needed? Just curious about how it works.

File: src/software/ai/evaluation/scoring/skills/feasibility_visitor.cpp [hunk 6]
```diff
@@ -32,44 +32,30 @@ void FeasibilityVisitor::visit(const KeepAwaySkill& skill)
...
     }
 
-    double shot_angle_viability =
-        normalizeValueToRange(best_shot->getOpenAngle().toDegrees(), 0.0, 60.0, 0.0, 1.0);
-    current_feasibility_ = shot_angle_viability;
+    double shot_origin_feasibility = rectangleSigmoid(
+        world_.field().enemyThird(), best_shot->getOrigin(), SIGMOID_WIDTH);
+
+    double shot_angle_feasibility =
+        normalizeValueToRange(best_shot->getOpenAngle().toDegrees(), 0.0,
```
Comment (Bvasilchikov) [inline src/software/ai/evaluation/scoring/skills/feasibility_visitor.cpp:55]: cool function, didn't know we had this.

File: src/software/ai/hl/stp/play/example_play.cpp
Comment (Bvasilchikov) [inline src/software/ai/hl/stp/play/example_play.cpp:1]: rip tests?

File: src/software/ai/hl/stp/play/play.cpp [hunk 7]
```diff
@@ -111,7 +111,7 @@ std::unique_ptr<TbotsProto::PrimitiveSet> Play::get(
     obstacle_list.Clear();
     path_visualization.Clear();
 
-    tactic_robot_id_assignment.clear();
+    std::map<std::shared_ptr<const Tactic>, RobotId> current_tactic_robot_id_assignment;
```
Comment (Bvasilchikov) [inline src/software/ai/hl/stp/play/play.cpp:114]: shouldn't `RobotId` be the key for a map, or am I mixing up how syntax works...

Comment (Bvasilchikov) [reply src/software/ai/hl/stp/play/play.cpp:114]: you would need to look into performance of swapping it or you already have - using tactics as keys implies that there is a constant set of tactics that doesn't change throughout an entire match, and the `RobotId` values are changing based on the situation.

File: src/software/ai/hl/stp/skill/pivot_kick/pivot_kick_skill_fsm.cpp [hunk 8]
```diff
@@ -24,13 +24,14 @@ void PivotKickSkillFSM::kickBall(const Update& event)
 
 bool PivotKickSkillFSM::lostPossession(const Update& event)
```
Comment (Bvasilchikov) [inline src/software/ai/hl/stp/skill/pivot_kick/pivot_kick_skill_fsm.cpp]: I would call this `robotLostLocalBallTraction` or something similar to not confuse with the case when the team turns over possession of the ball to the enemy team and they have control over it. maybe we can discuss different levels of control/possession as it will probably be useful in `PlayMonitor` too.

File: src/software/ai/hl/stp/skill/shoot/shoot_skill_fsm.cpp [hunk 9]
```diff
@@ -1,19 +1,60 @@
...
     {
-        kick_target = best_shot->getPointToShootAt();
+        best_shot_ = (*event.common.strategy)->getBestShot(event.common.robot);
+
+        if (!best_shot_)
+        {
+            // Default kick origin and target if no shot found
+            Point kick_origin = event.common.world_ptr->ball().position();
+            Point kick_target = event.common.world_ptr->field().enemyGoalCenter();
+            best_shot_        = Shot(kick_origin, kick_target, Angle::zero());
```
Comment (Bvasilchikov) [inline src/software/ai/hl/stp/skill/shoot/shoot_skill_fsm.cpp:36]: suggestion: try a chip shot or do a 50/50 coin-flip for kick/chip to confuse enemy team.

Comment (Bvasilchikov) [reply src/software/ai/hl/stp/skill/shoot/shoot_skill_fsm.cpp:36]: suggestion: maybe add a `MAX_SHOT_SEARCH_ATTEMPTS` and set to 3. then add a small delay interval to potentially look for a shot and wait before trying again.

File: src/software/ai/hl/stp/strategy/strategy_impl.hpp [hunk 10]
```diff
@@ -194,15 +195,14 @@ Pass StrategyImpl<PitchDivision, ZoneEnum>::getBestCommittedPass()
 template <class PitchDivision, class ZoneEnum>
 std::optional<Shot> StrategyImpl<PitchDivision, ZoneEnum>::getBestShot(const Robot& robot)
 {
-    if (robot_to_best_shot_.contains(robot.id()))
+    if (!best_shot_)
```
Comment (Bvasilchikov) [inline src/software/ai/hl/stp/strategy/strategy_impl.hpp]: seeing this pattern quite a bit for checking for best_shot existence - maybe we can fall back on the second or third best for these cases, relating to the suggesting of sorting the sample passes.

File: src/software/ai/evaluation/calc_best_shot.cpp (hunk 3 above)
Comment (Bvasilchikov) [reply src/software/ai/evaluation/calc_best_shot.cpp]: now that I think of it, it would be really helpful to accumulate & return a list of all sampled shots that have a direct line that scores, and sort by some "feasibility" depending on the enemy defence positioning.

**Review by Bvasilchikov — APPROVED** [review]
Awesome work!!

I added some gameplay related suggestions you can try, and hyped to discuss this during meetings!

File: src/software/ai/evaluation/calc_best_shot.cpp (hunk 3 above)
Comment (williamckha) [reply src/software/ai/evaluation/calc_best_shot.cpp]: ~~`std::transform` is O(n) and will dominate over O(nlogn)~~ not sure what I was thinking, you're right, O(nlogn) > O(n). Time complexity doesn't really matter here because n is always smallish. I'll convert this to using std algorithms anyways for stylistic purposes.

What potential use cases are you thinking of for returning a list of all feasible sampled shots? Ideally we always want to take the most feasible shot, so I'm not sure if there's a practical reason why we would need the 2nd, 3rd, 4th, etc. best shot

File: src/software/ai/evaluation/scoring/skills/feasibility_visitor.cpp (hunk 5 above)
Comment (williamckha) [reply src/software/ai/evaluation/scoring/skills/feasibility_visitor.cpp:36]: Not sure 😄 I have PassSkill disabled locally so I forgot I changed this. It will be rewritten in the future

File: src/software/ai/hl/stp/play/play.cpp (hunk 7 above)
Comment (williamckha) [reply src/software/ai/hl/stp/play/play.cpp:114]: Using the tactic as the key is how the original authors of `Play` designed the tactic assignment map. Its not a great choice but I guess it works. Performance doesn't matter here since the number of tactics is small. 

I don't think using tactics as keys implies that there is a constant set of tactics. This function always gets a pointer to the same tactic instance while the tactic is executing (we only create new tactics when we switch plays or a tactic finishes). So by using the pointer as the key for the tactic instance, the tactic will always have the same key in the map while it's still running. Once the tactic finishes, it will be removed from the map (since we clear the map every tick and we've lost the pointer to the finished tactic). 

i.e. The keys in the map represent the tactic instances currently running.

File: src/software/ai/hl/stp/skill/shoot/shoot_skill_fsm.cpp (hunk 9 above)
Comment (williamckha) [reply src/software/ai/hl/stp/skill/shoot/shoot_skill_fsm.cpp:36]: We'll have a `ChipSkill`, so the SkillGraph will dictate whether we kick or chip

File: src/software/ai/hl/stp/strategy/strategy_impl.hpp (hunk 10 above)
Comment (williamckha) [reply src/software/ai/hl/stp/strategy/strategy_impl.hpp]: If best_shot is std::nullopt, that implies that no feasible shot exists (so there isn't a 2nd or 3rd possible shot to fall back on). e.g. maybe we are behind the net

File: src/software/ai/evaluation/calc_best_shot.cpp (hunk 2 above)
Comment (williamckha) [reply src/software/ai/evaluation/calc_best_shot.cpp]: i'll make these dynamic params

**Comment by williamckha** [comment]
Merged in upstream master, hence the large diff
